presto-main/src/main/java/com/facebook/presto/sql/planner/NodePartitioningManager.java

```diff
@@ -258,9 +258,24 @@ private ConnectorBucketNodeMap getConnectorBucketNodeMap(Session session, Partit
         return distribution.build();
     }
 
-    public List<Node> sortedNodes(Session session, ConnectorId connectorId)
+    public List<Node> getNodes(Session session, ConnectorId connectorId)
     {
-        List<InternalNode> nodes = nodeScheduler.createNodeSelector(session, connectorId).allNodes();
-        return nodes.stream().sorted(comparing(InternalNode::getNodeIdentifier)).collect(toImmutableList());
+        // Nodes returned by the node selector are already sorted based on nodeIdentifier. No need to sort again
+        List<InternalNode> allNodes = nodeScheduler.createNodeSelector(session, connectorId).allNodes();
+
+        ImmutableList.Builder<Node> nodeBuilder = ImmutableList.builder();
+        for (int i = 0; i < allNodes.size(); i++) {
+            InternalNode node = allNodes.get(i);
+            if (node.getNodeStatus() == DEAD) {
+                // Replace dead nodes with the first alive node to the right of it in the sorted node list
+                int index = i + 1;
+                while (node.getNodeStatus() == DEAD) {
```

Comment:
This could break with overflow index.Example if the last node is dead.

Reply:
Good catch ! Will fix it.